Amazon Kineses 1.0 (#1096)
mulesoft/docs-connectors: amazon-kinesis/1.0/modules/ROOT/pages/amazon-kinesis-connector-reference.adoc

= Amazon Amazon Kinesis Data Streams Connector 1.0 Reference - Mule 4

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

== Configurations
---
[[Config]]
=== Config

The configuration for the connector. You can use this configuration for both the producer and consumer.

==== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
|Name | String | The name for this configuration. Connectors reference the configuration with this name. | | x
| The connection types to provide to this configuration. | | x

> | Default Stream Name a| String |  The default stream name. |  |

Reviewer: didn't we decided to drop it due to ConfigOverride limitations? @adam-lipka

| Connection Timeout a| Number |  The amount of time to wait when initially establishing the TCP connection between the connector and Amazon Kinesis server before throwing an exception if the connection fails |  30 |
| Connection Timeout Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  The time unit for the *Connection Timeout* field. |  SECONDS |
| Response Timeout a| Number |  The maximum time to wait between the time an HTTP request is sent and when a response is received |  30 |
| Response Timeout Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  Time unit for the *Response Timeout* value. |  SECONDS |
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  |
|===

== Supported Operations
* <<Checkpoint>>
* <<PutRecord>>

==== Associated Sources
* <<Listener>>

== Operations

[[Checkpoint]]
== Checkpoint
`<kinesis:checkpoint>`

> The operation checkpoint current position of a listener into DynamoDB table. Upon failover, the connector starts fetching records after the checkpointed position.

Reviewer: add: at the end: "it doesn't modify the current position of the listener"

Uses the Amazon Kinesis Client Library (KCL) to listen from a specified point in a stream.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Output Mime Type a| String |  The MIME type of the payload that this operation outputs. |  |
| Output Encoding a| String |  The encoding of the payload that this operation outputs. |  |
| Config Ref a| ConfigurationProvider |  The name of the configuration to use to execute this component |  | x
| Primary Node Only a| Boolean |  Whether this listener executes on the primary node only when running in a cluster |  |
| On Capacity Overload a| Enumeration, one of:

** WAIT
** DROP |  Strategy that Mule applies when the flow receives more messages than it can handle |  WAIT |
| Redelivery Policy a| <<RedeliveryPolicy>> |  Policy for processing the redelivery of a message |  |
| Stream Name a| String |  The stream name. |  |
| Application Name a| String |  The DynamoDB table that holds data about current stream positions. If the table doesn't exist, then the connector creates a new one at the initial position. |  |

> | Consumer ID a| String |  The unique ID of the consumer used for shard lease management. This value must be unique, or an `InvalidArgumentException` message is thrown. This value is used for the *Seek* operation and for Amazon DynamoDB lease management. The KCL LeaseManager object uses the ID to identify which shard is reserved for the consumer identified by the consumer ID. The Consumer ID must be unique for the application name. |  | x

Reviewer: consumer ID parameter is not present anymore @adam-lipka

= Amazon Amazon Kinesis Data Streams Connector 1.0 Reference - Mule 4

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

== Configurations
---
[[Config]]
=== Config

The configuration for the connector. You can use this configuration for both the producer and consumer.

==== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
|Name | String | The name for this configuration. Connectors reference the configuration with this name. | | x
| Connection a| <<Config_Connection, Kinesis Connection>> | The connection types to provide to this configuration. | | x

> | Default Stream Name a| String |  The default stream name. |  |
> | Default Application Name a| String |  The default application name. |  |

Reviewer 2: This was removed.
```suggestion
```

| Record TTL a| Number | Time limit for data records to be put. Records that are not successfully put within this limit fail. When setting this value, take into consideration the fact that data records still incur network latency after they leave the Amazon Kinesis Producer Library (KPL). |  30 |
| Record TTL Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  The time unit for the *Record TTL* value. |  SECONDS |
| Thread Pool Size a| Number |  Maximum number of threads with which to configure the native process thread pool. The default value, 0, enables the KPL process to choose the size of the thread pool. |  0 |
|===

==== Connection Types
[[Config_Connection]]
===== Kinesis Connection

> Connects to AWS Kinesis through the use of a user-specified access and secret key and, optionally, a role. These credentials securely sign requests to AWS services.

Reviewer 2: ```suggestion
Connects to AWS Kinesis through the use of a user-specified access and secret key and, optionally, a role. These credentials securely sign requests sent to AWS services.
```

====== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required

> | Proxy Configuration a| <<ProxyConfiguration>> |  Configuration for outbound connections through a proxy. |  |

Reviewer 2: The proxy configuration works only for listener and the checkpoint operation. Currently we have in our javadoc the following description:
```suggestion
| Proxy Configuration a| <<ProxyConfiguration>> |  Proxy connection settings used for outbound connections. This setting is only applied for the listener (source) and the Checkpoint operation, which means it will have no effect on the Put Record operation. |  |
```

> | TLS Configuration a| <<Tls>> |  Defines a configuration for TLS, which can be used from both the client and server sides to secure communication for the Mule app. When using the HTTPS protocol, the HTTP communication is secured using TLS or SSL. If HTTPS is configured as the protocol, then, at a minimum, the user must configure the keystore in the `tls:context` child element of the `listener-config`.  |  HTTP|

Reviewer 2: Same as with Proxy, the Put Record operation does not support it. Only listener and the Checkpoint operation does.
```suggestion
| TLS Configuration a| <<Tls>> |  Protocol to use for communication. Valid values are HTTP and HTTPS. Default value is HTTP. When using HTTPS the HTTP communication is going to be secured using TLS / SSL. If HTTPS was configured as protocol then the user needs to configure at least the keystore in the tls:context child element of this listener-config. It's noteworthy that this settings is only applied to the listener (source) and the Checkpoint operation. This will have no effect on the Put Record operation.  |  HTTP|
```

| Response Timeout a| Number |  The maximum time to wait between the time an HTTP request is sent and when a response is received |  30 |
| Response Timeout Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  Time unit for the *Response Timeout* value. |  SECONDS |
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  |
|===

== Supported Operations
* <<Checkpoint>>
* <<PutRecord>>

==== Associated Sources
* <<Listener>>

== Operations

[[Checkpoint]]
== Checkpoint
`<kinesis:checkpoint>`

The operation checkpoint position of a listener into the Amazon DynamoDB table. Upon failover, the connector starts fetching records that are located after the check pointed position. The *Checkpoint* operation does not modify the current position of the listener.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x

> | Application Name a| String |  Name of the DynamoDB table that holds data about the current stream positions. |  |

Reviewer 2: Stream name parameter is missing:
```suggestion
| Stream Name a| String |  The stream name. |  | x
| Application Name a| String |  Name of the DynamoDB table that holds data about the current stream positions. |  | x
```

| Shard ID a| String |  The shard identifier. |  |
| Sequence Number a| String |  A sequence number at which to checkpoint in the shard. |  |
| Subsequence Number a| Number |  A subsequence number at which to checkpoint within the shard. |  |
| Config Ref a| ConfigurationProvider |  The name of the configuration to use to execute this component |  | *x*
|===

=== For Configurations

* <<Config>>

=== Throws

* KINESIS:ILLEGAL_STATE
* KINESIS:INVALID_ARGUMENT

> * KINESIS:TIMEOUT

Reviewer 2: Error type was added:
```suggestion
* KINESIS:PROVISIONED_THROUGHPUT_EXCEEDED
* KINESIS:TIMEOUT
```

| Message Identifier a| <<RedeliveryPolicyMessageIdentifier>> | The strategy used to identify the messages. |  |
| Object Store a| ObjectStore | The object store that stores the redelivery counter for each message. |  |
|===

[[RedeliveryPolicyMessageIdentifier]]
=== Redelivery Policy Message Identifier

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Use Secure Hash a| Boolean | Whether to use a secure hash algorithm to identify a redelivered message. |  |
| Id Expression a| String | One or more expressions to use to determine when a message was redelivered. This property can be set only if the value of the *Use Secure Hash* field is `false`. |  |
|===

[[ShardConfig]]
=== Shard Config

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Shard Id a| String | The shard identifier. |  | x
| Shard Initial Position a| <<ShardInitialPosition>> | The shard initial position. |  | x
|===

[[ShardInitialPosition]]
=== Shard Initial Position

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | At Sequence Number a| String |  |  |

Reviewer 3: at SN: `This option ensures that listeners will listen from a specific sequence number.`
After SN: `The option ensures that listeners will listen after a specific sequence number.`
timestamp: `This option ensures that listeners will listen from a specific date and time.`
absolute position: `This option ensures that the listener will listen from the latest or oldest records.`

= Amazon Amazon Kinesis Data Streams Connector 1.0 Reference - Mule 4

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

== Configurations
---
[[Config]]
=== Config

The configuration for the connector. You can use this configuration for both the producer and consumer.

==== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
|Name | String | The name for this configuration. Connectors reference the configuration with this name. | | x
| Connection a| <<Config_Connection, Kinesis Connection>> | The connection types to provide to this configuration. | | x
| Default Stream Name a| String |  The default stream name. |  |
| Default Application Name a| String |  The default application name. |  |

> | Aggregation Enabled a| Boolean |  When `true`, multiple user data records are packed into a single Amazon Kinesis data  record. |  true |

Reviewer 4: ```suggestion
| Aggregation Enabled a| Boolean |  When set to `true`, packs multiple user data records into a single Amazon Kinesis data record. |  true |
```

| Response Timeout a| Number |  The maximum total time that can elapse between the start of an HTTP request and when the responses are received. If this time is exceeded, the request times out.|  30 |
| Request Timeout Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  A time unit that qualifies the *Response Timeout* value |  SECONDS |

> | Record Max Buffered Time a| Number |  Maximum amount of time for which an Amazon Kinesis data record spends being buffered before it is sent. Records can be sent earlier than this value, depending on the other buffering limits. |  3 |

Reviewer 4: ```suggestion
| Record Max Buffered Time a| Number |  Maximum amount of time that an Amazon Kinesis data record spends buffering before it is sent. Records can be sent earlier than this value, depending on other buffering limits. |  3 |
```

| Min Connections a| Number |  Minimum number of connections to keep open to the backend. In most cases, there is no need to increase value. |  1 |
| Rate Limit a| Number |  Limits the number of data records per second and number of bytes per second sent to a shard. This limit can reduce bandwidth and CPU cycle wastage from sending requests that will fail from throttling. The default value, 150%, enables a single producer instance to completely saturate the allowance for a shard. To reduce throttling errors rather than completely saturate the shard, consider reducing this setting. |  150% |
| Record TTL a| Number | Time limit for data records to be put. Records that are not successfully put within this limit fail. When setting this value, take into consideration the fact that data records still incur network latency after they leave the Amazon Kinesis Producer Library (KPL). |  30 |
| Record TTL Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  The time unit for the *Record TTL* value. |  SECONDS |

> | Thread Pool Size a| Number |  Maximum number of threads with which to configure the native process thread pool. The default value, 0, allows the KPL process to choose the size of the thread pool. |  0 |

Reviewer 4: ```suggestion
| Thread Pool Size a| Number |  Maximum number of threads to configure the native process thread pool. The default value, 0, enables the KPL process to choose the size of the thread pool. |  0 |
```

| Connection Timeout a| Number |  The amount of time to wait when initially establishing the TCP connection between the connector and Amazon Kinesis server before throwing an exception if the connection fails |  30 |
| Connection Timeout Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  The time unit for the *Connection Timeout* field. |  SECONDS |
| Response Timeout a| Number |  The maximum time to wait between the time an HTTP request is sent and when a response is received |  30 |
| Response Timeout Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  Time unit for the *Response Timeout* value. |  SECONDS |
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  |
|===

== Supported Operations
* <<Checkpoint>>
* <<PutRecord>>

==== Associated Sources
* <<Listener>>

== Operations

[[Checkpoint]]
== Checkpoint
`<kinesis:checkpoint>`

> The operation checkpoint position of a listener into the Amazon DynamoDB table. Upon failover, the connector starts fetching records that are located after the check pointed position. The *Checkpoint* does not modify the current position of the listener.

Reviewer 4: ```suggestion
The operation checkpoint position of a listener into the Amazon DynamoDB table. Upon failover, the connector starts fetching records that are located after the check pointed position. The *Checkpoint*  operation does not modify the current position of the listener.
```

Reviewer 4: I'm not sure what the first sentence is trying to say:
`The operation checkpoint position of a listener into the Amazon DynamoDB table`
Does it mean _The *Checkpoint* operation positions a a listener into the Amazon DynamoDB table_? Please confirm with devs.

| Shard ID a| String |  The shard identifier. |  |
| Sequence Number a| String |  A sequence number at which to checkpoint in the shard. |  |
| Subsequence Number a| Number |  A subsequence number at which to checkpoint within the shard. |  |
| Config Ref a| ConfigurationProvider |  The name of the configuration to use to execute this component |  | *x*
|===

=== For Configurations

* <<Config>>

=== Throws

* KINESIS:ILLEGAL_STATE
* KINESIS:INVALID_ARGUMENT
* KINESIS:TIMEOUT
* KINESIS:UNAUTHORIZED

[[PutRecord]]
== Put Record
`<kinesis:put-record>`

Puts a data record into an Amazon Kinesis data stream.

=== Parameters

[%header%autowidth.spread]
|===
| Name | Type | Description | Default Value | Required
| Configuration | String | The name of the configuration to use. | | x
| Stream Name a| String |  The stream name |  |
| Partition Key a| String |  The partition key. If not provided, the connector generates a UUID for this value. |  |
| Explicit Hash Key a| String |  Overrides the Amazon Kinesis Data Streams explicitHashKey value, which is normally computed with using MD5 function of the data record partition key. |  |

> | Data a| Any |  The content to put in the Amazon Kinesis data stream. This can be upt o 1 MB in size. |  #[payload] |

Reviewer 4: ```suggestion
| Data a| Any |  The content to put in the Amazon Kinesis data stream. Up to 1 MB in size. |  #[payload] |
```

** NO_OP_SHARD_PRIORITIZATION | Logic used to prioritize or filter the shards before their execution. The options are:

* PARENTS_FIRST_SHARD_PRIORITIZATION
Prioritizes parent shards first and limits the number of shards that are available for initialization based on their depth (*Max Depth* field).
* NO_OP_SHARD_PRIORITIZATION
Returns the original list of shards without any modifications. |  NO_OP_SHARD_PRIORITIZATION |
| Max Depth a| Number |  Used by the PARENTS_FIRST_SHARD_PRIORITIZATION shard prioritization option. Any shard that is deeper than this value is excluded from processing.  |  |
| Read Capacity Units a| Number |  The maximum number of strongly consistent reads consumed per second before Amazon DynamoDB returns a ThrottlingException. |  10 |
| Write Capacity Units a| Number |  The maximum number of writes consumed per second before Amazon DynamoDB returns a ThrottlingException. |  10 |
| Failover Time a| Number |  Duration of a lease for a worker. |  10 |
| Failover Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS |  Time unit for the *Failover Time* field. |  SECONDS |

> | Checkpoint On Complete a| Boolean |  If `true`, the listener checkpoints the current reading position to Amazon DynamoDB after data records are processed by the owning flow, without taking into account whether the processing finished successfully. |  true |

Reviewer 4: ```suggestion
| Checkpoint On Complete a| Boolean |  If `true`, the listener checkpoints the current reading position to Amazon DynamoDB after data records are processed by the flow, without taking into account whether the processing finished successfully or not |  true |
```

| Non Proxy Hosts a| Array of String | A list of hosts against which the proxy should not be used. |  |
| Ntlm Domain a| String | The domain to authenticate against the proxy. |  |
|===

[[Tls]]
=== TLS

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Enabled Protocols a| String | A comma-separated list of protocols enabled for this context. |  |
| Enabled Cipher Suites a| String | A comma-separated list of cipher suites enabled for this context. |  |
| Trust Store a| <<TrustStore>> |  |  |
| Key Store a| <<KeyStore>> |  |  |
| Revocation Check a| * <<StandardRevocationCheck>>
* <<CustomOcspResponder>>
* <<CrlFile>> |  |  |
|===

[[TrustStore]]
=== Trust Store

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Path a| String | The location (which is resolved relative to the current classpath and file system, if possible) of the truststore. |  |
| Password a| String | The password used to protect the truststore. |  |
| Type a| String | The type of store used. |  |
| Algorithm a| String | The algorithm used by the truststore. |  |

> | Insecure a| Boolean | If true, no certificate validations are performed, which makes connections vulnerable to attacks. Use at your own risk. |  |

Reviewer 4: ```suggestion
| Insecure a| Boolean | If `true`, no certificate validations are performed, which makes connections vulnerable to attacks. Use at your own risk. |  |
```

> | Insecure a| Boolean | If true, no certificate validations are performed, which makes connections vulnerable to attacks. Use at your own risk. |  |

Reviewer 4: ```suggestion
| Insecure a| Boolean | If set to `true`, no certificate validations are performed, which makes connections vulnerable to attacks. Use at your own risk. |  |
```

|===

[[KeyStore]]
=== Key Store

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Path a| String | The location (which is resolved relative to the current classpath and file system, if possible) of the keystore. |  |
| Type a| String | The type of store used. |  |
| Alias a| String | When the keystore contains many private keys, this attribute indicates the alias of the key that should be used. If not defined, the first key in the file is used by default. |  |
| Key Password a| String | The password used to protect the private key. |  |
| Password a| String | The password used to protect the keystore. |  |
| Algorithm a| String | The algorithm used by the keystore. |  |
|===

[[StandardRevocationCheck]]
=== Standard Revocation Check

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Only End Entities a| Boolean | Verifies the last element of the certificate chain only. |  |
| Prefer Crls a| Boolean | Tries CRL instead of OCSP first. |  |

> | No Fallback a| Boolean | Does not use the secondary checking method (the one not selected before). |  |

Reviewer 4: ```suggestion
| No Fallback a| Boolean | Does not use the secondary checking method. |  |
```

Reviewer 4: Not sure if _the one not selected before_ refers to **Prefer Crls**

|===

[[ReconnectForever]]
=== Reconnect Forever

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Frequency a| Number | How often in milliseconds to reconnect |  |
|===

[[RecordAttributes]]
=== Record Attributes

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Aggregated a| Boolean |  |  |
| Approximate Arrival Timestamp a| String |  |  |
| Encryption Type a| String |  |  |
| Partition Key a| String |  |  |
| Partition Key a| String |  |  |
| Sequence Number a| String |  |  |
| Shard Id a| String |  |  |
| Sub Sequence Number a| Number |  |  |
|===

[[RedeliveryPolicy]]
=== Redelivery Policy

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Max Redelivery Count a| Number | The maximum number of times a message can be redelivered and processed unsuccessfully before triggering a process-failed message |  |
| Message Digest Algorithm a| String | The secure hashing algorithm to use. |  SHA-256|
| Message Identifier a| <<RedeliveryPolicyMessageIdentifier>> | The strategy used to identify the messages. |  |

> | Object Store a| ObjectStore | The object store where the redelivery counter for each message is stored. |  |

Reviewer 4: ```suggestion
| Object Store a| ObjectStore | The object store to store the redelivery counter for each message. |  |
```
